TranscriptAnnotator.pm

```diff
@@ -0,0 +1,260 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+=head1 CONTACT
+
+ Ensembl <http://www.ensembl.org/info/about/contact/index.html>
+
+=cut
+
+=head1 NAME
+
+ TranscriptAnnotator.pm
+
+=head1 SYNOPSIS
+
+ mv TranscriptAnnotator.pm ~/.vep/Plugins
+ ./vep -i variations.vcf --plugin TranscriptAnnotator,file=/path/to/file.txt.gz
+
+=head1 DESCRIPTION
+
+ A VEP plugin that annotates variant-transcript pairs based on a given file:
+
+   --plugin TranscriptAnnotator,file=${HOME}/file.tsv.gz
+
+ Example of a valid tab-separated annotation file (before bgzip and tabix):
+
+   #Chrom  Pos       Ref  Alt  Transcript       SIFT_score  SIFT_pred    Comment
+   11      436154    A    G    NM_001347882.2   0.03        Deleterious  Bad
+   11      1887471   C	  T    ENST00000421485  0.86        Tolerated    Good
+
+ Parameters:
+
+   file:     Tabix-indexed file to parse. Must contain variant location
+             (chromosome, position, reference allele, alternative allele) and
+             transcript ID as the first 5 columns. Accepted transcript IDs
+             include those from Ensembl and RefSeq.
+   cols:     Colon-delimited list with names of the columns to append. Column
+             names are based on the last header line. By default, all columns
+              (except the first 5) are appended.
+   prefix:   String to prefix the name of appended columns (default: basename of
+             the filename without extensions). Set to 0 to avoid any prefix.
+   trim:     Trim whitespaces from both ends of each column (default: 1).
+
+ The tabix and bgzip utilities must be installed in your path to read the
+ tabix-indexed annotation file: check https://github.com/samtools/htslib.git for
+ installation instructions.
+
+=cut
+
+package TranscriptAnnotator;
+
+use strict;
+use warnings;
+
+use File::Basename;
+
+use Bio::EnsEMBL::Variation::Utils::Sequence qw(get_matched_variant_alleles);
+
+use Bio::EnsEMBL::Variation::Utils::BaseVepTabixPlugin;
+use base qw(Bio::EnsEMBL::Variation::Utils::BaseVepTabixPlugin);
+
+sub _basename_without_ext {
+  my $file = basename shift;
+  $file =~ s/\.[^.]*?(.gz)?$//g;
+  return $file;
+}
+
+sub _trim_whitespaces {
+  my $text = shift;
+  $text =~ s/^\s+|\s+$//g;
+  return $text;
+};
+
+sub _get_colnames {
+  my $self = shift;
+
+  # Open file header
+  open IN, "tabix -f -h " . $self->{_files}[0] . " 1:1-1 |"
+    or die "ERROR: cannot open tabix file for " . $self->{_files}[0];
+
+  # Get last line
+  my $last;
+  $last = $_ while <IN>;
+  $last =~ s/(^#|\n$)//g;
+  close IN;
+
+  # Parse column names from header
+  my @cols = split /\t/, $last;
+  @cols = map { _trim_whitespaces $_ } @cols if $self->{trim};
+  @cols = splice @cols, 5;
+  return @cols;
```

Comment:
Here it should check if the first 5 columns are the correct columns:
chromosome, position, reference allele, alternative allele, transcript id

I tested with a file missing the alt allele and the plugin didn't fail but it also didn't return anything. In this case, the plugin should fail.

Reply:
What is the best option to check for this? Using mandatory column names?

Reply:
Yes, mandatory names could avoid missing values. 
The plugin already expects the columns to be fixed anyway. You can add to the description the specific names the first 5 columns have to have: `chr, pos, ref, alt, transcript_id` or `'chromosome, position, ref allele, alt allele, transcript_id'`?

Reply:
Okay, changed to check if the first five columns start with `chr, pos, ref, alt, tr`, which allows users to use either one of your suggestions!